### PR TITLE
convert MDS checkpoint to Hf Llama model

### PR DESCRIPTION
**Pull request by vksastry**

### CODE REVIEW

File: ALCF/mds_to_hf.py [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+# This script currently only takes care of tp=1. Takes a AuroraGPT Llama model trained with Megatron-DeepSpeed and converts to LLamaCausalForLM architecture from HuggingFace. 
+
+import argparse
+import torch
+import pdb
+import os
+from transformers import LlamaConfig, LlamaForCausalLM, LlamaTokenizer
+
+def repeat_kv_wt(x,np):
+    return torch.repeat_interleave(x, dim=0, repeats=np)
```
Comment (saforem2) [inline ALCF/mds_to_hf.py:12]: this seems unused ??
(commented out on 64 + 66)

File: ALCF/mds_to_hf.py [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+    hidden_size = mds_model['args'].__dict__['hidden_size']
+    kv_dim = mds_model['args'].__dict__['kv_channels'] * mds_model['args'].__dict__['num_key_value_heads']
+    kv_groups = mds_model['args'].__dict__['num_attention_heads'] // mds_model['args'].__dict__['num_key_value_heads']
+    for layer_i in range(Updated_Llama_config.__dict__['num_hidden_layers']):
+        # SELF ATTENTION layers.
+        # get the q, k, v weights separately. Keeping k and v at the GQA head dim, since the transformers/models/llama/modelling_utils will take care of it. 
+        fused_qkv = mds_model['module']['language_model']['encoder'][f"layers.{layer_i}.self_attention.query_key_value.weight"]
+        state_dict[f"model.layers.{layer_i}.self_attn.q_proj.weight"] = fused_qkv[0:hidden_size]
+        state_dict[f"model.layers.{layer_i}.self_attn.k_proj.weight"] = fused_qkv[hidden_size:hidden_size+kv_dim]
+        #state_dict[f"model.layers.{layer_i}.self_attn.k_proj.weight"] = repeat_kv_wt(fused_qkv[hidden_size:hidden_size+kv_dim], kv_groups)
```
Comment (saforem2) [inline ALCF/mds_to_hf.py:64]:
```suggestion
```

File: ALCF/mds_to_hf.py [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+        state_dict[f"model.layers.{layer_i}.self_attn.v_proj.weight"] = fused_qkv[hidden_size+kv_dim:hidden_size+2*kv_dim]
+        #state_dict[f"model.layers.{layer_i}.self_attn.v_proj.weight"] = repeat_kv_wt(fused_qkv[hidden_size+kv_dim:hidden_size+2*kv_dim],kv_groups)
```
Comment (saforem2) [inline ALCF/mds_to_hf.py:66]:
```suggestion
```

File: ALCF/mds_to_hf.py [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+        state_dict[f"model.layers.{layer_i}.self_attn.rotary_emb.inv_freq"] = inv_freq
+
+    # Get the non-encoder layer weights. 
+    state_dict["model.embed_tokens.weight"] = mds_model['module']['language_model']['embedding']['word_embeddings']['weight']
+    state_dict["model.norm.weight"] = mds_model['module']['language_model']['encoder']['final_layernorm.weight']
+    state_dict["lm_head.weight"] = mds_model['module']['language_model']['output_layer']['weight']
+    
+    # Save the model in the hf output path. 
+    torch.save(state_dict, os.path.join(args.output_dir,"pytorch_model.bin"))    
+
```
Comment (saforem2) [inline ALCF/mds_to_hf.py:89]:
```suggestion
```

**Comment by vksastry** [comment]
I thought probably we may need this if we change the architecture in future. But if it is hurting, I can remove them.

**Comment by saforem2** [comment]
🤷🏻‍♂️ leave 'em in there if they're useful

**Pull request merged**